docs/design-notes/per-partition-rate-limit.md

```diff
@@ -0,0 +1,142 @@
+
+_NOTE_: this is an overload protection mechanism and may not be used to reliably
+enforce limits in some situations. Due to Scylla's distributed nature,
+the actual number of accepted requests depends on the cluster and driver
+configuration and may be larger by a factor of RF (keyspace's replication
+factor). It is recommended to set the limit to a value an order of magnitude
+larger than the maximum expected per-partition throughput.
+
+## Usage
+
+### Server-side configuration
+
+Per-partition limits are set separately for reads and writes, on a per-table
+basis. Limits can be set with the `per_partition_rate_limit` extension when
+CREATE'ing or ALTER'ing a table using a schema extension:
+
+```cql
+ALTER TABLE ks.tbl WITH per_partition_rate_limit = {
+    'max_reads_per_second': 123,
+    'max_writes_per_second': 456
+};
+```
+
+Both `max_reads_per_second` and `max_writes_per_second` are optional - omitting
+one of them means "no limit" for that type of operation.
+
+### Driver response
+
+Rejected operations are reported as an ERROR response to the driver.
+If the driver supports it, the response contains a scylla-specific error code
+indicating that the operation was rejected. For more details about the error
+code, see the [Rate limit error](./protocol-extensions.md#Rate%20limit%20error)
+section in the `protocol-extensions.md` doc.
+
+If the driver doesn't support the new error code, the `Config_error` code
+is returned instead. The code was chosen in order for the retry policies
+of the drivers not to retry the requests and instead propagate them directly
+to the users.
+
+## How it works
+
+Accounting related to tracking per-partition limits is done by replicas.
+Each replica keeps a map of counters which are identified by a combination
+of (token, table, operation type). When the replica accounts an operation,
+it increments the relevant counter. All counters are halved every second.
+
+Depending on whether the coordinator is a replica or not, the flow is
+a bit different. Here, "coordinator == replica" requirement also means
+that the operation is handled on the correct shard.
+
+Only reads and writes explicitly issued by the user are counted to the limit.
+Read repair, hints, batch replay, CDC preimage query and internal system queries
+are _not_ counted to the limit.
+
+Paxos and counters are not covered in current implemenation.
+
+### Coordinator is not a replica
+
+Coordinator generates a random number from range `[0, 1)` with uniform
+distribution and sends it to replicas along with the operation request.
+Each replica accounts the operation and then calculates a rejection threshold
+based on the local counter value. If the number received from the coordinator
+is above the threshold, the operation is rejected.
+
+The assumption is that all replicas will converge to similar counter values.
+Most of the time they will agree on the decision and not much work
+will be wasted due to some replicas accepting and other rejecting.
```

Comment:
Can you rerun the test you ran with previous series to see how many replicas reject the write ?

Reply:
I added a feature which synchronizes the rate limiter tick with the system time. The results of the test will now be highly variable on how well the nodes' clocks are synchronized.

I re-ran the test described [here](https://github.com/scylladb/scylla/pull/9810#discussion_r879643131). All nodes were running on the same machine, and additionally I disabled the "coordinator optimization" so that it does not influence the results. In 99.8% cases replicas were in agreement.

We can estimate the worst case. Basing the calculations on [this](https://github.com/scylladb/scylla/pull/9810#issuecomment-1155071509) and assuming a steady request rate of `V`, replicas will be in agreement if the random number sent by coordinator is _outside_ the range `[P(2V), P(V)]`. The chance of _disagreement_ is at most:

```
P(V) - P(2V)
= (L/V) * (1 / ln 2 - 1 / (2 * ln 2))
= (L/V) * (1 / (2 * ln 2))
~= (L/V) * 0.7213475204444817
```

So, `0.72 * L` out of `V` requests per second may have disagreeing replicas.

Reply:
What is 'L' her? Also will reboot of a node affect the result since it loses it local view of the rate limit?

Reply:
Yes, temporarily. The per-partition counters are not preserved and are reset to 0 on reboot. Until the new node "heats up" and its counter value stabilizes between V and 2V, it will serve more requests. The counter converges exponentially, from what I noticed it takes about 5 seconds - then, assuming that the clocks are still synchronized after reboot, the agreement rate should be the same.

Reply:
So suppose we have a scenario when all nodes were rate limiting and two were rebooted. Now those two start to accept but one still rejects and thus requests succeed, but we are in a situation we try to avoid when hints are stored and immediately replayed.

Reply:
Yes, the newly restarted nodes will accumulate some hints for a short while until they start agreeing, then they will be replayed to the old node. I'm not sure how we can avoid it when replicas independently decide whether to reject or not.

/cc @avikivity @psarna

Reply:
We could try solving it by gathering information about the counters from replicas, caching them and decaying on the coordinator, like mentioned [here](https://github.com/scylladb/scylla/pull/9810#discussion_r879498072). The question is whether we want to do it and if we want to do it now. If it is implemented, then some of the current parts of the solution will become obsolete (e.g. typed exceptions over RPC).

Reply:
I think it's okay if we generate some noise on restarts.

Reply:
Back when we wrote HWLB we were very careful to avoid noise on restart. I wouldn't want to complicate the solution to remove the noise either though. IMO, which I already stated before, the cleanest solution would be to support topology aware drivers only since we pretty much require using them for best performance. This would have simplified the patch series greatly as well.